### PR TITLE
Unsigned arith corrected for JS

### PR DESCRIPTION
**Pull request by yglukhov**
Corrected unsigned arithmetics to reflect C behavior. Added a unittest to js category which previously would fail. Removed extra code. Enabled `$``(uint)`.

### CODE REVIEW

File: compiler/jsgen.nim [hunk 1]
```diff
@@ -336,90 +336,6 @@ const # magic checked op; magic unchecked op; checked op; unchecked op
     ["cstrToNimstr", "cstrToNimstr", "cstrToNimstr($1)", "cstrToNimstr($1)"],
     ["", "", "$1", "$1"]]
 
-  luaOps: TMagicOps = [
```
Comment (Araq) [inline compiler/jsgen.nim]: Aww, what about the Lua codegen?

File: lib/system.nim [hunk 2]
```diff
@@ -1754,11 +1754,12 @@ proc `$`*(x: int64): string {.magic: "Int64ToStr", noSideEffect.}
...
   ## converted to a decimal string.
 
-when not defined(nimscript):
-  when not defined(JS) and hasAlloc:
+when not defined(nimscript) and not defined(JS) and hasAlloc:
     proc `$` *(x: uint64): string {.noSideEffect.}
       ## The stringify operator for an unsigned integer argument. Returns `x`
       ## converted to a decimal string.
+else:
+    proc `$` *(x: uint32): string {.magic: "Int64ToStr", noSideEffect.}
```
Comment (Araq) [inline lib/system.nim]: This will bite us. Not sure what to do instead though.

File: compiler/jsgen.nim (hunk 1 above)
Comment (yglukhov) [inline compiler/jsgen.nim]: Well, I can put it back if it makes you feel better =)

File: lib/system.nim (hunk 2 above)
Comment (yglukhov) [inline lib/system.nim]: @Araq, could you elaborate please? At least if you're not gonna let this change in ;)

**Comment by yglukhov** [comment]
Corrected trimming. Ready to go.

**Comment by yglukhov** [comment]
Reverted $(uint64) change. Ready to go.

File: compiler/jsgen.nim [hunk 3]
```diff
@@ -1406,11 +1342,17 @@ proc genMagic(p: PProc, n: PNode, r: var TCompRes) =
     else:
       unaryExpr(p, n, r, "", "($1 != null ? ($1.length-1) : -1)")
   of mInc:
-    if optOverflowCheck notin p.options: binaryExpr(p, n, r, "", "$1 += $2")
-    else: binaryExpr(p, n, r, "addInt", "$1 = addInt($1, $2)")
+    if n.sons[1].typ.kind in tyUInt .. tyUInt64:
```
Comment (Araq) [inline compiler/jsgen.nim]: n[1].typ.skipTypes(abstractRange).kind in {...}

File: compiler/jsgen.nim [hunk 4]
```diff
@@ -1406,11 +1342,17 @@ proc genMagic(p: PProc, n: PNode, r: var TCompRes) =
...
+      binaryUintExpr(p, n, r, "+", true)
+    else:
+      if optOverflowCheck notin p.options: binaryExpr(p, n, r, "", "$1 += $2")
+      else: binaryExpr(p, n, r, "addInt", "$1 = addInt($1, $2)")
   of ast.mDec:
-    if optOverflowCheck notin p.options: binaryExpr(p, n, r, "", "$1 -= $2")
-    else: binaryExpr(p, n, r, "subInt", "$1 = subInt($1, $2)")
+    if n.sons[1].typ.kind in tyUInt .. tyUInt64:
```
Comment (Araq) [inline compiler/jsgen.nim]: Ditto.

**Comment by yglukhov** [comment]
Done. sorry, still not used to this skipType thing. should i do the same in binaryUintExpr?

**Comment by Araq** [comment]
Good question. Can't hurt.

**Comment by yglukhov** [comment]
Done =)

**Pull request merged**